mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -479,9 +492,16 @@ class Word2VecModel private[mllib] (
    */
   def findSynonyms(vector: Vector, num: Int): Array[(String, Double)] = {
     require(num > 0, "Number of similar words should > 0")
-    // TODO: optimize top-k
-    val fVector = vector.toArray.map(_.toFloat)
-    model.mapValues(vec => cosineSimilarity(fVector, vec))
+
+    val numWords = wordVectors.numRows
+    val cosineVec = Vectors.zeros(numWords).asInstanceOf[DenseVector]
+    BLAS.gemv(1.0, wordVectors, vector.asInstanceOf[DenseVector], 0.0, cosineVec)
+
+    // Need not divide with the norm of the given vector since it is constant.
+    val updatedCosines = indexedModel.map { case (_, ind) =>
```

Comment:
Here's where you're assuming something about the ordering of Map items.  Instead of this, I'd recommend just operating on cosineVec and wordVecNorms.  After finding the top word indices, you can look them up in the wordIndex (indexedModel).  I'd also use a while loop instead of a map since it should be faster.

I also wonder if we should store the word index in both directions (since we need both in the 2 findSynonyms methods):
* ```wordList: Array[String]``` (words ordered by indices)
* ```wordIndex: Map[String, Int]``` (map from word to index)